opt_einsum/contract.py

```diff
@@ -805,6 +805,9 @@ def contract_expression(subscripts, *shapes, **kwargs):
             raise ValueError("'{}' should only be specified when calling a "
                              "`ContractExpression`, not when building it.".format(arg))
 
+    if not isinstance(subscripts, compat.strings):
```

Comment:
Any reason why we just do not call `parse_einsum_input` here? It would be good, if possible to always parse the input at a single location. I think current `parse_einsum_input` does not yet support non-array like inputs such as tuples, but this should easy to correct I believe.

Reply:
So, it's just because ``contract_expression`` routes through ``contract`` and all its machinery once the shapes and constants are processed and thus ``parse_einsum_input`` will be called again anyway. The only reason ``convert_interleaved_input`` is needed here (which will then not be called again), is to simplify the ``constants`` processing.

One option might be have ``contract`` accept a tuple of the parsed inputs, or a ``is_parsed`` option or something, so that ``contract_expression`` could do the parsing and call ``contract`` with the guarantee that it is already parsed.

Reply:
It might be good to break out the inner contract loops at some point to make the actual iterations more flexible.